proposals/0000-hypercore.md

```diff
@@ -0,0 +1,274 @@
+
+The depth of an tree node can be calculated by counting the number of trailing 1s a node has in binary notation.
+
+```
+5 in binary = 101 (one trailing 1)
+3 in binary = 011 (two trailing 1s)
+4 in binary = 100 (zero trailing 1s)
+```
+      
+1 is the parent of (0, 2), 5 is the parent of (4, 6), and 3 is the parent of (1, 5).
+
+If the number of leaf nodes is a power of 2 the flat tree will only have a single root. Otherwise it'll have more than one. As an example here is a tree with 6 leafs:
+
+```
+0
+  1
+2
+    3
+4
+  5
+6
+
+8
+  9
+10
+```
+      
+The roots spanning all the above leafs are 3 an 9. Throughout this document we'll use following tree terminology:
+
+ - `parent` - a node that has two children (odd numbered)
+ - `leaf` - a node with no children (even numbered)
+ - `sibling` - the other node with whom a node has a mutual parent
+ - `uncle` - a parent's sibling
+
+
+## Merkle Trees
+[merkle-trees]: #merkle-trees
+
+A merkle tree is a binary tree where every leaf is a hash of a data block and every parent is the hash of both of its children. Hypercore feeds are merkle trees encoded with "bin numbers" (see above).
+
+Let's look at an example. A feed containing four values would be mapped to the even numbers 0, 2, 4, and 6.
+
+```
+chunk0 -> 0
+chunk1 -> 2
+chunk2 -> 4
+chunk3 -> 6
+```
+
+Let `h(x)` be a hash function. Using flat-tree notation, the merkle tree spanning these data blocks looks like this:
+
+```
+0 = h(chunk0)
+1 = h(0 + 2)
+2 = h(chunk1)
+3 = h(1 + 5)
+4 = h(chunk2)
+5 = h(4 + 6)
+6 = h(chunk3)
+```
+
+In the resulting Merkle tree, the even and odd nodes store different information:
+
+- Evens - List of data hashes [chunk0, chunk1, chunk2, ...]
+- Odds - List of Merkle hashes (hashes of child even nodes) [hash0, hash1, hash2, ...]
+
+In a merkle tree, the "root node" hashes the entire data set. In this example of 4 chunks, node 3 hashes the entire data set. Therefore we only need to trust node 3 to verify all data. As entries are added to a Hypercore feed, the "active" root node will change.
+
+```
+0
+  1
+2
+    3 (root node)
+4
+  5
+6
+```
+
+It is possible for the in-order Merkle tree to have multiple roots at once. For example, let's expand our example dataset to contain six items. This will result in two roots:
+
+```
+0
+  1
+2
+    3 (root node)
+4
+  5
+6
+
+8
+  9 (root node)
+10
+```
+
+The nodes in this tree would be calculated as follows:
+
+```
+0 = h(chunk0)
+1 = h(0 + 2)
+2 = h(chunk1)
+3 = h(1 + 5)
+4 = h(chunk2)
+5 = h(4 + 6)
+6 = h(chunk3)
+
+8 = h(chunk4)
+9 = h(8 + 10)
+10 = h(chunk5)
+```
+
+In the Hypercore feed, we only want one active root. Therefore, when there are multiple roots we hash all the roots together again. We also include a big endian uint64 binary representation of the corresponding node index (TODO: why?). At most there will be `log2(number of data blocks)`.
+
+```
+root = h(uint64be(#9) + 9 + uint64be(#3) + 3)
```

Comment:
We encode the index also as that describes the structure of the tree not just the data. This might not be nessesary but my gut feeling is that you could do a tree preimage attack on it https://en.m.wikipedia.org/wiki/Merkle_tree#Second_preimage_attack. We already defend against the preimage using the hash 1 byte prefixes so this might not be nessesary here but it doesn't hurt

Reply:
It basically very strongly encodes the length of the feed in the hash